// @farcaster/snap/dist/react-native/components/snap-stack.js
import { jsx as _jsx } from "react/jsx-runtime";
import { StyleSheet, View } from "react-native";
const VGAP = {
    none: 0,
    sm: 8,
    md: 16,
    lg: 24,
};
const HGAP = {
    none: 0,
    sm: 4,
    md: 8,
    lg: 12,
};
const JUSTIFY = {
    start: "flex-start",
    center: "center",
    end: "flex-end",
    between: "space-between",
    around: "space-around",
};
export function SnapStack({ element: { props }, children, }) {
    const direction = String(props.direction ?? "vertical");
    const rawGap = props.gap;
    const isHorizontal = direction === "horizontal";
    const gapMap = isHorizontal ? HGAP : VGAP;
    const gap = typeof rawGap === "number"
        ? rawGap
        : typeof rawGap === "string" && rawGap in gapMap
            ? gapMap[rawGap]
            : isHorizontal ? HGAP.md : VGAP.md;
    const justify = props.justify ? JUSTIFY[String(props.justify)] : undefined;
    return (_jsx(View, { style: [
            styles.stack,
            isHorizontal ? styles.horizontal : undefined,
            { gap },
            justify ? { justifyContent: justify } : undefined,
        ], children: children }));
}
const styles = StyleSheet.create({
    stack: {
        width: "100%",
    },
    horizontal: {
        flexDirection: "row",
        alignItems: "center",
        flexWrap: "wrap",
    },
});
