// @farcaster/snap/dist/react-native/components/snap-progress.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyleSheet, Text, View } from "react-native";
import { useSnapPalette } from "../use-snap-palette.js";
import { useSnapTheme } from "../theme.js";
export function SnapProgress({ element: { props }, }) {
    const { accentHex } = useSnapPalette();
    const { colors } = useSnapTheme();
    const value = Number(props.value ?? 0);
    const max = Math.max(1, Number(props.max ?? 100));
    const percent = Math.min(100, Math.max(0, (value / max) * 100));
    const label = props.label != null ? String(props.label) : null;
    return (_jsxs(View, { style: styles.wrap, children: [label ? (_jsx(Text, { style: [styles.label, { color: colors.textSecondary }], children: label })) : null, _jsx(View, { style: [styles.track, { backgroundColor: colors.muted }], children: _jsx(View, { style: [styles.fill, { width: `${percent}%`, backgroundColor: accentHex }] }) })] }));
}
const styles = StyleSheet.create({
    wrap: { width: "100%", gap: 4 },
    label: { fontSize: 13, lineHeight: 18 },
    track: {
        height: 10,
        borderRadius: 9999,
        overflow: "hidden",
    },
    fill: {
        height: "100%",
        borderRadius: 9999,
    },
});
