// @farcaster/snap/dist/schemas.js
import { z } from "zod";
import { EFFECT_VALUES, SUPPORTED_SPEC_VERSIONS } from "./constants.js";
import { DEFAULT_THEME_ACCENT, PALETTE_COLOR_VALUES } from "./colors.js";
// ─── Theme ─────────────────────────────────────────────
const themeAccentSchema = z.enum(PALETTE_COLOR_VALUES, {
    message: `accent must be a palette color: ${PALETTE_COLOR_VALUES.join(", ")}`,
});
const themeSchema = z
    .object({
    accent: themeAccentSchema.default(DEFAULT_THEME_ACCENT),
})
    .strict();
// ─── Snap response ─────────────────────────────────────
// `ui` is a json-render Spec — validated by the catalog at runtime,
// typed here via the json-render Spec type.
export const snapResponseSchema = z
    .object({
    version: z.enum(SUPPORTED_SPEC_VERSIONS),
    theme: themeSchema.optional().default({ accent: DEFAULT_THEME_ACCENT }),
    effects: z.array(z.enum(EFFECT_VALUES)).optional(),
    ui: z.custom((val) => val != null &&
        typeof val === "object" &&
        "root" in val &&
        "elements" in val, { message: "ui must be a json-render Spec with root and elements" }),
})
    .strict();
// ─── POST payload ──────────────────────────────────────
const postInputValueSchema = z.union([
    z.string(),
    z.number(),
    z.boolean(),
    z.array(z.string()),
]);
export const payloadSchema = z
    .object({
    fid: z.number().int().nonnegative(),
    inputs: z.record(z.string(), postInputValueSchema).default({}),
    timestamp: z.number().int(),
    nonce: z.string(),
    audience: z.string(),
})
    .strip();
export const ACTION_TYPE_GET = "get";
export const ACTION_TYPE_POST = "post";
const snapGetActionSchema = z.object({
    type: z.literal(ACTION_TYPE_GET),
});
const snapPostActionSchema = payloadSchema.extend({
    type: z.literal(ACTION_TYPE_POST),
});
export const snapActionSchema = z.discriminatedUnion("type", [
    snapGetActionSchema,
    snapPostActionSchema,
]);
