// @farcaster/snap/dist/react/components/stack.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { cn } from "@neynar/ui/utils";
const VGAP = {
    none: "gap-0",
    sm: "gap-2",
    md: "gap-4",
    lg: "gap-6",
};
const HGAP = {
    none: "gap-0",
    sm: "gap-1",
    md: "gap-2",
    lg: "gap-3",
};
const JUSTIFY = {
    start: "justify-start",
    center: "justify-center",
    end: "justify-end",
    between: "justify-between",
    around: "justify-around",
};
export function SnapStack({ element: { props }, children, }) {
    const direction = String(props.direction ?? "vertical");
    const gapKey = String(props.gap ?? "md");
    const isHorizontal = direction === "horizontal";
    const gap = isHorizontal
        ? (HGAP[gapKey] ?? "gap-2")
        : (VGAP[gapKey] ?? "gap-4");
    const justify = props.justify ? JUSTIFY[String(props.justify)] : undefined;
    return (_jsx("div", { className: cn("flex w-full", isHorizontal ? "flex-row items-center flex-wrap" : "flex-col", gap, justify), children: children }));
}
