// @farcaster/snap/dist/react/v2/snap-view.js
"use client";
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useMemo } from "react";
import { validateSnapResponse } from "../../validator.js";
import { SnapViewCore } from "../snap-view-core.js";
const SNAP_MAX_HEIGHT = 500;
const SNAP_WARNING_HEIGHT = 700;
// ─── Default validation error fallback ────────────────
function SnapValidationFallback({ appearance, message, }) {
    const isDark = appearance === "dark";
    return (_jsx("div", { style: {
            width: "100%",
            padding: 16,
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
            color: isDark ? "rgba(255,255,255,0.5)" : "rgba(0,0,0,0.4)",
            fontSize: 14,
        }, children: _jsx("span", { children: message ? `Unable to render snap: ${message}` : "Unable to render snap" }) }));
}
// ─── SnapViewV2 ──────────────────────────────────────
export function SnapViewV2({ snap, handlers, loading = false, appearance = "dark", onValidationError, validationErrorFallback, }) {
    const validation = useMemo(() => validateSnapResponse(snap), [snap]);
    const valid = validation.valid;
    const validationMessage = validation.issues[0]?.message;
    useEffect(() => {
        if (!valid) {
            if (onValidationError) {
                onValidationError(validation);
            }
            else {
                // eslint-disable-next-line no-console
                console.warn("[Snap] validation issues:", validation.issues);
            }
        }
    }, [valid, validation, onValidationError]);
    if (!valid) {
        if (validationErrorFallback === null)
            return null;
        return _jsx(_Fragment, { children: validationErrorFallback ?? _jsx(SnapValidationFallback, { appearance: appearance, message: validationMessage }) });
    }
    return (_jsx(SnapViewCore, { snap: snap, handlers: handlers, loading: loading, appearance: appearance }));
}
// ─── SnapCardV2 ──────────────────────────────────────
export function SnapCardV2({ snap, handlers, loading = false, appearance = "dark", maxWidth = 480, showOverflowWarning = false, onValidationError, validationErrorFallback, actionError, plain = false, }) {
    const maxHeight = showOverflowWarning ? SNAP_WARNING_HEIGHT : SNAP_MAX_HEIGHT;
    const isDark = appearance === "dark";
    const bg = isDark ? "rgba(0,0,0,0.85)" : "rgba(255,255,255,0.9)";
    const borderColor = isDark ? "rgba(255,255,255,0.1)" : "rgba(0,0,0,0.1)";
    const surfaceBg = isDark ? "rgba(255,255,255,0.05)" : "rgba(0,0,0,0.02)";
    return (_jsxs(_Fragment, { children: [_jsxs("div", { style: {
                    position: "relative",
                    width: "100%",
                    maxWidth,
                    maxHeight,
                    overflow: "hidden",
                    ...(plain ? {} : {
                        borderRadius: 16,
                        border: `1px solid ${borderColor}`,
                        backgroundColor: surfaceBg,
                    }),
                }, children: [_jsx("div", { style: plain ? undefined : { padding: 16 }, children: _jsx(SnapViewV2, { snap: snap, handlers: handlers, loading: loading, appearance: appearance, onValidationError: onValidationError, validationErrorFallback: validationErrorFallback }) }), showOverflowWarning && (_jsxs("div", { style: {
                            position: "absolute",
                            top: SNAP_MAX_HEIGHT,
                            left: 0,
                            right: 0,
                            bottom: 0,
                            pointerEvents: "none",
                            zIndex: 10,
                        }, children: [_jsx("div", { style: { borderTop: "1px dashed rgba(255,100,100,0.6)", position: "relative" }, children: _jsxs("span", { style: {
                                        position: "absolute",
                                        top: -10,
                                        right: 0,
                                        fontSize: 10,
                                        fontFamily: "ui-monospace, SFMono-Regular, Menlo, monospace",
                                        color: "rgba(255,100,100,0.7)",
                                        background: bg,
                                        padding: "1px 4px",
                                        borderRadius: 3,
                                    }, children: [SNAP_MAX_HEIGHT, "px"] }) }), _jsx("div", { style: {
                                    height: "100%",
                                    background: "repeating-linear-gradient(-45deg, transparent, transparent 8px, rgba(255,100,100,0.06) 8px, rgba(255,100,100,0.06) 16px)",
                                } })] }))] }), actionError && (_jsx("div", { style: {
                    maxWidth,
                    padding: "8px 12px",
                    fontSize: 13,
                    color: appearance === "dark"
                        ? "rgba(255,100,100,0.9)"
                        : "rgba(200,0,0,0.8)",
                }, children: actionError }))] }));
}
