// @farcaster/snap/dist/react/v1/snap-view.js
"use client";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useRef, useState } from "react";
import { SnapViewCore } from "../snap-view-core.js";
const SNAP_MAX_HEIGHT = 500;
export function SnapViewV1({ snap, handlers, loading = false, appearance = "dark", }) {
    return (_jsx(SnapViewCore, { snap: snap, handlers: handlers, loading: loading, appearance: appearance }));
}
export function SnapCardV1({ snap, handlers, loading = false, appearance = "dark", maxWidth = 480, actionError, plain = false, }) {
    const isDark = appearance === "dark";
    const borderColor = isDark ? "rgba(255,255,255,0.1)" : "rgba(0,0,0,0.1)";
    const surfaceBg = isDark ? "rgba(255,255,255,0.05)" : "rgba(0,0,0,0.02)";
    const toggleBg = isDark ? "rgba(255,255,255,0.06)" : "rgba(0,0,0,0.05)";
    const toggleBgHover = isDark
        ? "rgba(255,255,255,0.1)"
        : "rgba(0,0,0,0.08)";
    const toggleText = isDark ? "rgba(255,255,255,0.82)" : "rgba(0,0,0,0.72)";
    const contentRef = useRef(null);
    const [isExpandable, setIsExpandable] = useState(false);
    const [isExpanded, setIsExpanded] = useState(false);
    useEffect(() => {
        setIsExpanded(false);
    }, [snap]);
    useEffect(() => {
        const node = contentRef.current;
        if (!node)
            return;
        const measure = () => {
            setIsExpandable(node.scrollHeight > SNAP_MAX_HEIGHT + 1);
        };
        measure();
        if (typeof ResizeObserver === "undefined")
            return;
        const observer = new ResizeObserver(() => {
            measure();
        });
        observer.observe(node);
        return () => observer.disconnect();
    }, [snap, plain]);
    useEffect(() => {
        if (!isExpandable) {
            setIsExpanded(false);
        }
    }, [isExpandable]);
    const isClipped = isExpandable && !isExpanded;
    return (_jsxs("div", { style: {
            position: "relative",
            width: "100%",
            maxWidth,
            overflow: "hidden",
            ...(plain ? {} : {
                borderRadius: 16,
                border: `1px solid ${borderColor}`,
                backgroundColor: surfaceBg,
            }),
        }, children: [_jsx("div", { style: isClipped
                    ? {
                        maxHeight: SNAP_MAX_HEIGHT,
                        overflow: "hidden",
                    }
                    : undefined, children: _jsx("div", { ref: contentRef, style: plain ? undefined : { padding: 16 }, children: _jsx(SnapViewV1, { snap: snap, handlers: handlers, loading: loading, appearance: appearance }) }) }), isExpandable ? (_jsx("div", { style: {
                    display: "flex",
                    justifyContent: "center",
                    padding: plain ? "8px 0 0" : "10px 16px 12px",
                    ...(plain
                        ? {}
                        : { borderTop: `1px solid ${borderColor}` }),
                }, children: _jsx("button", { type: "button", "aria-expanded": isExpanded, onClick: () => setIsExpanded((value) => !value), style: {
                        appearance: "none",
                        border: "none",
                        borderRadius: 9999,
                        backgroundColor: toggleBg,
                        color: toggleText,
                        padding: "6px 10px",
                        fontSize: 13,
                        lineHeight: "18px",
                        fontWeight: 600,
                        cursor: "pointer",
                    }, onMouseEnter: (event) => {
                        event.currentTarget.style.backgroundColor = toggleBgHover;
                    }, onMouseLeave: (event) => {
                        event.currentTarget.style.backgroundColor = toggleBg;
                    }, children: isExpanded ? "Show less" : "Show more" }) })) : null, actionError && (_jsx("div", { style: {
                    padding: "8px 12px",
                    fontSize: 13,
                    color: appearance === "dark"
                        ? "rgba(255,100,100,0.9)"
                        : "rgba(200,0,0,0.8)",
                }, children: actionError }))] }));
}
