// @farcaster/snap/dist/server/verify.js
import { compact, decode, decodePayload as jfsDecodePayload, encodePayload as jfsEncodePayload, verify, } from "@farcaster/jfs";
import { hexToBytes } from "viem";
import { DEFAULT_SNAP_HUB_HTTP_BASE_URL, getActiveEd25519SignerKeysFromHubHttp, } from "./hubs.js";
export async function verifyJFSRequestBody(requestBody, options = {}) {
    let compactJfs;
    try {
        compactJfs = compact({
            header: requestBody.header,
            payload: requestBody.payload,
            signature: requestBody.signature,
        });
    }
    catch (error) {
        return {
            valid: false,
            error: error instanceof Error ? error : new Error(String(error)),
        };
    }
    let decoded;
    try {
        decoded = decode(compactJfs);
    }
    catch (error) {
        return {
            valid: false,
            error: error instanceof Error ? error : new Error(String(error)),
        };
    }
    try {
        await verify({ data: compactJfs, strict: true, keyTypes: ["app_key"] });
    }
    catch (error) {
        return {
            valid: false,
            error: error instanceof Error ? error : new Error(String(error)),
        };
    }
    const { header, payload } = decoded;
    const keys = await getActiveEd25519SignerKeysFromHubHttp(options.hubHttpBaseUrl ?? DEFAULT_SNAP_HUB_HTTP_BASE_URL, header.fid);
    if (!keys.ok) {
        return {
            valid: false,
            error: new Error(keys.message),
        };
    }
    let headerKeyBytes;
    try {
        headerKeyBytes = hexToBytes(header.key);
    }
    catch {
        return {
            valid: false,
            error: new Error("invalid JFS header key encoding"),
        };
    }
    if (headerKeyBytes.length !== 32) {
        return {
            valid: false,
            error: new Error("JFS app_key public key must be 32 bytes"),
        };
    }
    const matched = keys.signers.some((k) => bytesEqual(k.publicKey, headerKeyBytes));
    if (!matched) {
        return {
            valid: false,
            error: new Error("active hub signer list does not include JFS header key"),
        };
    }
    return {
        valid: true,
        data: payload,
    };
}
export function decodePayload(payload) {
    return jfsDecodePayload(payload);
}
export function encodePayload(payload) {
    return jfsEncodePayload(payload);
}
function bytesEqual(a, b) {
    if (a.length !== b.length)
        return false;
    let diff = 0;
    for (let i = 0; i < a.length; i++)
        diff |= a[i] ^ b[i];
    return diff === 0;
}
